src/plugins/layout/shadowdom.ts

```diff
@@ -272,25 +208,93 @@ export class ShadowDom {
       this.removeAllClasses(next);
     }
 
-    let removed = this.removedNodes.childNodes;
-    for (let i = 0; i < removed.length; i++) {
-      let next = removed[i] as IShadowDomNode;
-      if (!this.hasClass(next, NewNodeClassName)) {
-        summary.removedNodes.push(next);
-      }
+    let removedNodes = this.removedNodes.childNodes;
+    for (let i = 0; i < removedNodes.length; i++) {
+      traverseNodeTree(removedNodes[i], (shadowNode: IShadowDomNode) => {
+        if (this.hasClass(shadowNode, NewNodeClassName)) {
+          delete shadowNode.node[NodeIndex];
+        } else if (this.hasClass(shadowNode, MovedNodeClassName)) {
+          summary.removedNodes.push(shadowNode);
+        }
+      });
     }
 
     return summary;
   }
 
+  public createDomIndexJson(): object {
+
+    function writeIndex(node: Node, json: object) {
```

Comment:
can't you find a better type for this? Something to indicate it is a mapping from numbers to maps of numbers?